pkg/k8s/watchers/cilium_envoy_config.go

```diff
@@ -0,0 +1,160 @@
+//  Copyright 2020 Authors of Cilium
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package watchers
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (k *K8sWatcher) ciliumEnvoyConfigInit(ciliumNPClient *k8s.K8sCiliumClient) {
+	cecStore := cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
+
+	cecController := informer.NewInformerWithStore(
+		cache.NewListWatchFromClient(ciliumNPClient.CiliumV2alpha1().RESTClient(),
+			cilium_v2alpha1.CECPluralName, v1.NamespaceAll, fields.Everything()),
+		&cilium_v2alpha1.CiliumEnvoyConfig{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricCreate, valid, equal) }()
+				if cec := k8s.ObjToCEC(obj); cec != nil {
+					valid = true
+					err := k.addCiliumEnvoyConfig(cec)
+					k.K8sEventProcessed(metricCEC, metricCreate, err == nil)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricUpdate, valid, equal) }()
+
+				if oldCEC := k8s.ObjToCEC(oldObj); oldCEC != nil {
+					if newCEC := k8s.ObjToCEC(newObj); newCEC != nil {
+						valid = true
+						if newCEC.DeepEqual(oldCEC) {
+							equal = true
+							return
+						}
+						err := k.updateCiliumEnvoyConfig(oldCEC, newCEC)
+						k.K8sEventProcessed(metricCEC, metricUpdate, err == nil)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricDelete, valid, equal) }()
+				cec := k8s.ObjToCEC(obj)
+				if cec == nil {
+					return
+				}
+				valid = true
+				err := k.deleteCiliumEnvoyConfig(cec)
+				k.K8sEventProcessed(metricCEC, metricDelete, err == nil)
+			},
```

Comment:
We have an open issue to refactor any new additions to the k8s watcher to use the `subscriber` pattern. See https://github.com/cilium/cilium/issues/15471. The issue gives an example to follow. One suggestion: the intention of the `subscriber` package is to be extended whenever there's a new resource to be notified for. This PR adds the CEC object, so you can make your changes to `pkg/k8s/watchers/subscriber/cec.go` (needs to be created). Let me know if I can clarify / help with this.

Reply:
I'll take a stab at this, was not aware of [#15471](https://github.com/cilium/cilium/issues/15471).